timemachine/cpp/src/context.cu

```diff
@@ -47,6 +51,26 @@ Context::~Context() {
     gpuErrchk(cudaFree(d_sum_storage_));
 };
 
+void Context::_verify_box(cudaStream_t stream) {
+    // If there are no nonbonded potentials, nothing to check.
+    if (nonbonded_pots_.size() == 0) {
+        return;
+    }
+    std::vector<double> h_box(9);
+    gpuErrchk(cudaMemcpyAsync(&h_box[0], d_box_t_, 9 * sizeof(*d_box_t_), cudaMemcpyDeviceToHost, stream));
+    gpuErrchk(cudaStreamSynchronize(stream));
+    for (auto boundpot : nonbonded_pots_) {
+        double cutoff = get_nonbonded_all_pair_cutoff_with_padding(boundpot->potential);
+        double db_cutoff = 2 * cutoff;
+        for (int i = 0; i < 3; i++) {
+            if (h_box[i * 3 + i] < db_cutoff) {
+                throw std::runtime_error(
+                    "cutoff with padding is more than half of the box width, neighborlist is no longer reliable");
+            }
+        }
+    }
```

Comment:
I wonder if we can move some of this logic back into the kernel GPU code for the neighborlist, since CUDA code can now assert:

https://stackoverflow.com/questions/50755717/triggering-a-runtime-error-within-a-cuda-kernel

Reply:
Based on the documentation (https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#assertion) it looks like that will return a cudaError code which we will then call exit() inside of cudaErrchk.

Seems like an interesting idea, but would require more careful usage to avoid exiting which I find a bit less usable when working with jupyter notebooks.

Reply:
Hm - yeah. I guess we should probably clean-up cudaErrchk to throw first to be consistent. I would much prefer the box checks etc. to be done in the GPU code if possible (to keep the scope local to neighborlist code) so it doesn't pollute nblist code.